---
title: "01_qc_normalization"
author: "Chiara Caprioli"
date: "April 6th 2023"
output:
  html_document:
    code_folding: hide
---

**Aims:**
1. import cellranger count matrices and create data object (merging all samples)
2. remove doublets
3. remove cell with low gene and transcript counts (adaptive thresholds method)
4. remove lowly expressed genes
5. create merged Seurat object 
6. log normalization
7. cell cycle analysis
8. set batch metadata

## Set up
```{r, setup, include=FALSE}
# set working directory
knitr::opts_knit$set(root.dir = '/Users/ieo4874/Desktop/working/SCMseq/')
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, echo=F, message=F, warning=F}
#load packages
suppressPackageStartupMessages({
 library(tidyverse)
 library(Seurat)
 library(scran)
 library(scater)
 library(patchwork)
 library(viridis)
 library(ggforce)
 library(gghalves)
 library(ggridges)
 library(scDblFinder)
 library(SeuratObject)
})
```

```{r}
# set paths
path_main <- '/Users/ieo4874/Desktop/working/SCMseq/'
path_data <- paste0(path_main, 'data/')
path_results <- paste0(path_main, 'results_and_plots/qc_pp/')
```

```{r}
# samples
samples <- c('AML4', 'AML5', 'AML6', 'sAML1',
             'AML2', 'AML3', 'sAML16',
             'hBM1', 'hBM2', 'hBM3')
```

```{r, warning=FALSE, message=FALSE}
#set colors 
colors <- list()

colors$sample <- setNames(
  c("#FFEDA0", "#FEB24C", "#FC4E2A", "#BD0026", 
    "#66C2A5", "#3288BD", "#5E4FA2", 
    "#BDBDBD", "#737373", "#252525"),  
  samples
)

colors$SRSF2_genotype <- setNames(
  c("#FF9932FF", "#405D61FF", '#F0F0F0'),
  c("SRSF2 mutated", "SRSF2 unmutated", 'healthy BM')
)

colors$cell_cycle <- setNames(
  c("#45aaf2", "#f1c40f", "#e74c3c"),
  c("G1", "S", "G2M")
)

colors$discrete <- c(
  '#046C9AFF', '#D69C4EFF', '#899DA4FF', '#C93312FF', '#D8B70AFF',
  '#02401BFF', '#A2A475FF', '#81A88DFF', '#972D15FF', '#CADDE1FF',
  '#7B906FFF', '#174D79FF',  '#898989FF', '#D4CDB1FF', '#708090FF',
  '#6CA9C3FF', '#3A3533FF', '#000E33FF', '#800000FF',  '#175E78FF')
```

## Helper functions
```{r, warning=FALSE, message=FALSE}
niceFormat <- function(number) {
  formatC(number, format = "f", big.mark = ",", digits = 0)
}

load10xData <- function(
  sample_name,
  path,
  max_number_of_cells = NULL
) {

  ## read transcript count matrix
  transcript_counts <- list.files(
      path,
      recursive = TRUE,
      full.names = TRUE
    ) %>%
    .[grep(., pattern = 'mtx')] %>%
    Matrix::readMM()

  ## read cell barcodes and assign as column names of transcript count matrix
  colnames(transcript_counts) <- list.files(
      path,
      recursive = TRUE,
      full.names = TRUE
    ) %>%
    .[grep(., pattern = 'barcodes')] %>%
    .[grep(., pattern = 'tsv')] %>%
    read_tsv(col_names = FALSE, col_types = cols()) %>%
    pull(1) %>%
    gsub(., pattern = '-[0-9]{1,2}', replacement = '') %>%
    paste0(., '-', sample_name)

  ## read gene names
  gene_names <- list.files(
      path,
      recursive = TRUE,
      full.names = TRUE
    ) %>%
    .[grep(., pattern = 'genes|features')] %>%
    read_tsv(col_names = FALSE, col_types = cols()) %>%
    pull(ifelse(ncol(.) > 1, 2, 1))

  ## if necessary, merge counts from different transcripts of the same gene
  if ( any(duplicated(gene_names)) == TRUE ) {

    ## get names of genes with multiple entries
    duplicated_gene_names <- unique(gene_names[which(duplicated(gene_names))])

    ## log message
    message(
      glue::glue(
        "{format(Sys.time(), '[%Y-%m-%d %H:%M:%S]')} Summing up counts for ",
        "{length(duplicated_gene_names)} genes with multiple entries."
      )
    )

    ## go through genes with multiple entries
    for ( i in duplicated_gene_names ) {

      ## extract transcripts counts for current gene
      tmp_counts <- transcript_counts[which(gene_names == i),]

      ## make sure at least 2 rows are present
      if ( nrow(tmp_counts) >= 2 ) {

        ## sum up counts in each cell
        tmp_counts_sum <- Matrix::colSums(tmp_counts)

        ## remove original counts from transcript matrix and the gene name from
        ## the list of gene names
        transcript_counts <- transcript_counts[-c(which(gene_names == i)),]
        gene_names <- gene_names[-c(which(gene_names == i))]

        ## add summed counts to end of transcript count matrix and current gene
        ## name to end of gene names
        transcript_counts <- rbind(transcript_counts, tmp_counts_sum)
        gene_names <- c(gene_names, i)
      }
    }
  }

  ## assign gene names as row names
  rownames(transcript_counts) <- gene_names

  ## down-sample cells if more cells than `max_number_of_cells` are present in
  ## count matrix
  if (
    !is.null(max_number_of_cells) &&
    is.numeric(max_number_of_cells) &&
    max_number_of_cells < ncol(transcript_counts)
  ) {
    temp_cells_to_keep <- base::sample(1:ncol(transcript_counts), max_number_of_cells)
    transcript_counts <- transcript_counts[,temp_cells_to_keep]
  }

  ##
  message(
    glue::glue(
      "{format(Sys.time(), '[%Y-%m-%d %H:%M:%S]')} Loaded counts for sample ",
      "{sample_name} with {niceFormat(ncol(transcript_counts))} cells and ",
      "{niceFormat(nrow(transcript_counts))} genes."
    )
  )

  ##
  return(transcript_counts)
}

sameGeneNames <- function(counts) {

  ## create place holder for gene names from each count matrix
  gene_names <- list()

  ## go through count matrices
  for ( i in names(counts) ) {

    ## extract gene names
    gene_names[[i]] <- rownames(counts[[i]])
  }

  ## check if all gene names are the same (using the first matrix as a reference
  ## for the others)
  return(all(sapply(gene_names, FUN = identical, gene_names[[1]])))
}

averagenCountnFeature <- function(cells) {
  output <- tibble(
    'sample' = character(),
    'mean(nCount)' = numeric(),
    'median(nCount)' = numeric(),
    'mean(nFeature)' = numeric(),
    'median(nFeature)' = numeric()
  )
  for ( i in levels(cells$sample) ) {
    tmp <- tibble(
      'sample' = i,
      'mean(nCount)' = cells %>% filter(sample == i) %>% pull(nCount) %>% mean(),
      'median(nCount)' = cells %>% filter(sample == i) %>% pull(nCount) %>% median(),
      'mean(nFeature)' = cells %>% filter(sample == i) %>% pull(nFeature) %>% mean(),
      'median(nFeature)' = cells %>% filter(sample == i) %>% pull(nFeature) %>% median()
    )
    output <- bind_rows(output, tmp)
  }
  return(output)
}
```

## Load samples
```{r}
#sample_names <- list.dirs(paste0(path_main, 'matrices'), recursive = FALSE) %>% basename()

transcripts <- list(
  raw = list()
)

for ( i in samples ) {
  transcripts$raw[[i]] <- load10xData(
    i, paste0(path_main, 'matrices/', i, '/'),
    max_number_of_cells = NULL
  )
}
```

## Check gene names
```{r}
sameGeneNames(transcripts$raw) # should be TRUE
```

```{r}
# IN CASE OF NON IDENTICAL GENES


#for ( i in samples ) {
#  gene_names <- rownames(transcripts$raw[[i]])
#  transcripts$raw[[i]] <- as.data.frame(as.matrix(transcripts$raw[[i]]))
#  transcripts$raw[[i]] <- transcripts$raw[[i]] %>%
#    mutate(gene = gene_names) %>%
#    select(gene, everything())
#} # most likely results in memory issues
#
#transcripts$raw$merged <- 
#  full_join(transcripts$raw$A, transcripts$raw$B, by = "gene") %>%
#  full_join(., transcripts$raw$E, by = "gene") 
#  
#transcripts$raw$merged[ is.na(transcripts$raw$merged) ] <- 0

# check for which samples genes are not identical
#all_gene_names <- c()
#for ( i in samples ) {
#  gene_names <- rownames(transcripts$raw[[i]])
#  all_gene_names[[i]] <- gene_names
#}
#
#identical(names(all_gene_names), samples) # TRUE
#
#for (i in all_gene_names) {
#  print( length(i))
#}
#
#identical(all_gene_names$AML4, all_gene_names$AML5) # TRUE
#identical(all_gene_names$AML4, all_gene_names$sAML1) # FALSE
#identical(all_gene_names$AML4, all_gene_names$AML1) # TRUE
#identical(all_gene_names$AML4, all_gene_names$AML2) # TRUE
#identical(all_gene_names$AML4, all_gene_names$AML3) # TRUE
#identical(all_gene_names$AML4, all_gene_names$hBM1) # TRUE
#identical(all_gene_names$AML4, all_gene_names$hBM2) # TRUE
#identical(all_gene_names$AML4, all_gene_names$hBM3) # TRUE
#
#add_names <- rownames(transcripts$raw$AML4)[which(!rownames(transcripts$raw$AML4) %in% rownames(transcripts$raw$sAML1))] 
#length(add_names) #4683
#
#add_ <- matrix(data = NA, 
#                nrow = length(add_names), 
#                ncol = ncol(transcripts$raw$AML4), dimnames = NULL)
#rownames(add_) <- add_names
#transcripts$raw$AML4 <- rbind(transcripts$raw$AML4, add_)
#dim(transcripts$raw$AML4)
```

```{r}
transcripts$raw$merged <- do.call(cbind, transcripts$raw)
transcripts$raw$merged[ is.na(transcripts$raw$merged) ] <- 0
dim(transcripts$raw$merged)
```

## Quality control
```{r}
cells <- tibble(
  cell = colnames(transcripts$raw$merged),
  sample = colnames(transcripts$raw$merged) %>%
    strsplit("-") %>%
    vapply(FUN.VALUE = character(1), `[`, 2) %>%
    factor(levels = samples),
  nCount = colSums(transcripts$raw$merged),
  nFeature = colSums(transcripts$raw$merged != 0)
)

DataFrame(cells)
```

### add column with mitochondrial genes (HUMAN)
```{r}
mitochondrial_genes_here <- read_tsv(paste0(path_data, "genes_mt_hg_name.tsv.gz"),
  col_names = FALSE
) %>%
  filter(X1 %in% rownames(transcripts$raw$merged)) %>%
  pull(X1)

cells$percent_MT <-
  Matrix::colSums(transcripts$raw$merged[mitochondrial_genes_here,]) /
  Matrix::colSums(transcripts$raw$merged)

DataFrame(cells)
```

### find and remove doublets
Doublets can only arise within a given sample or capture, and for this reason are better sought independently for each sample, which also speeds up the analysis. If the 'samples' argument is given, scDblFinder will use it to split the cells into samples/captures, and process each of them in parallel if the BPPARAM argument is given. 
```{r}
sce <- SingleCellExperiment(
    assays = list(counts = transcripts$raw$merged),
    rowData = transcripts$raw$merged@Dimnames[[1]],
    colData = transcripts$raw$merged@Dimnames[[2]],
    metadata = list(sample = cells$sample)
    )

sce <- scDblFinder(
  sce,
  samples = sce@metadata$sample,
  multiSampleMode = 'split',
  BPPARAM = BiocParallel::SerialParam()
)

cells$multiplet_class <- colData(sce)$scDblFinder.class

cells %>%
  filter(multiplet_class != 'singlet') %>%
  group_by(sample) %>%
  summarize(count = n()) %>%
  ggplot(aes(x = sample, y = count, fill = sample)) +
  geom_col(color = 'black') +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_continuous(name = 'Number of doublets', labels = scales::comma) +
  theme(
    axis.title.y = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = 'none'
  ) +
  coord_flip()

ggsave(paste0(path_results, 'plots/qc_number_of_doublets_by_sample.png'), height = 3, width = 6
)
```

```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
p1 <- ggplot(cells, aes(x = multiplet_class, y = nCount, fill = multiplet_class)) +
  geom_violin(draw_quantiles = c(0.5), scale = 'area', trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$discrete) +
  scale_x_discrete(limits = rev(levels(cells$multiplet_class))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = 'Number of transcripts', subtitle = 'linear scale') +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = 'none'
  ) +
  coord_flip()
p2 <- ggplot(cells, aes(x = multiplet_class, y = nCount, fill = multiplet_class)) +
  geom_violin(draw_quantiles = c(0.5), scale = 'area', trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$discrete) +
  scale_x_discrete(limits = rev(levels(cells$multiplet_class))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = 'Number of transcripts', subtitle = 'log-scale') +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = 'none'
  ) +
  coord_flip()
p3 <- ggplot(cells, aes(x = multiplet_class, y = nFeature, fill = multiplet_class)) +
  geom_violin(draw_quantiles = c(0.5), scale = 'area', trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$discrete) +
  scale_x_discrete(limits = rev(levels(cells$multiplet_class))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = 'Number of expressed genes', subtitle = 'linear scale') +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = 'none'
  ) +
  coord_flip()
p4 <- ggplot(cells, aes(x = multiplet_class, y = nFeature, fill = multiplet_class)) +
  geom_violin(draw_quantiles = c(0.5), scale = 'area', trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$discrete) +
  scale_x_discrete(limits = rev(levels(cells$multiplet_class))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = 'Number of expressed genes', subtitle = 'log-scale') +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = 'none'
  ) +
  coord_flip()

ggsave(paste0(path_results, "plots/qc_nCount_nFeature_by_multiplet_class.png"),
  p1 + p3 + p2 + p4 + plot_layout(ncol = 2),
  height = 7, width = 10
)
```

```{r}
# save in case we want to check something back
saveRDS(cells, paste0(path_data, 'cells_unfiltered_with_doublets.rds'))
```

```{r}
# filter out doublets
cells <- filter(cells, multiplet_class == "singlet")
```

### find and remove low quality cells

#### check before filtering
```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}

p1 <- ggplot(cells, aes(x = sample, y = nCount, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = "Number of transcripts", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p2 <- ggplot(cells, aes(x = sample, y = nCount, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = "Number of transcripts", subtitle = "log-scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p3 <- ggplot(cells, aes(x = sample, y = nFeature, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = "Number of expressed genes", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p4 <- ggplot(cells, aes(x = sample, y = nFeature, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = "Number of expressed genes", subtitle = "log-scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p5 <- ggplot(cells, aes(x = sample, y = percent_MT, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "Percent MT transcripts", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()

ggsave(paste0(path_results,"plots/qc_histogram_nCount_nFeature_percentMT_before_filtering_1.png"),
  p1 + p3 + p5 +
    p2 + p4 + plot_layout(ncol = 3),
  height = 7, width = 10)

```

```{r}
p1 <- ggplot(cells, aes(nCount, fill = sample)) +
  geom_histogram(bins = 200) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_continuous(labels = scales::comma) +
  labs(title = 'Number of transcripts', subtitle = 'linear scale', y = 'Number of cells') +
  theme(legend.position = 'none', axis.title.x = element_blank())

p2 <- ggplot(cells, aes(nCount, fill = sample)) +
  geom_histogram(bins = 200) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_log10(labels = scales::comma) +
  labs(title = 'Number of transcripts', subtitle = 'log-scale', y = 'Number of cells') +
  theme(legend.position = 'none', axis.title.x = element_blank())

p3 <- ggplot(cells, aes(nFeature, fill = sample)) +
  geom_histogram(bins = 200) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_continuous(labels = scales::comma) +
  labs(title = 'Number of expressed genes', subtitle = 'linear scale', y = 'Number of cells') +
  theme(legend.position = 'none',axis.title.x = element_blank())

p4 <- ggplot(cells, aes(nFeature, fill = sample)) +
  geom_histogram(bins = 200) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_log10(labels = scales::comma) +
  labs(title = 'Number of expressed genes', subtitle = 'log-scale', y = 'Number of cells', fill = 'Sample') +
  theme(legend.position = 'none', axis.title.x = element_blank())

p5 <- ggplot(cells, aes(percent_MT, fill = sample)) +
  geom_histogram(bins = 200) +
  theme_bw() +
  scale_x_continuous(labels = scales::percent) +
  scale_fill_manual(name = 'Sample', values = colors$sample) +
  labs(title = 'Percent MT transcripts', subtitle = 'linear scale', y = 'Number of cells') +
  theme(legend.position = 'right', axis.title.x = element_blank(), legend.text = element_text(size = 8))

ggsave(paste0(path_results, 'plots/qc_histogram_ncount_nfeature_percentmt_before_filtering_3.png'),
  p1 + p3 + p5 +
  p2 + p4 + plot_layout(ncol = 3),
  height = 7, width = 10
)
```

#### Plot nFeature/nCount
```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}

cells %>% 
  ggplot(aes(nCount, nFeature, color = percent_MT)) +
  geom_point(size = 0.1) +
  scale_x_continuous(name = "Number of transcripts", labels = scales::comma) +
  scale_y_continuous(name = "Number of expressed genes", labels = scales::comma) +
  theme_bw() +
  scale_color_viridis(
    name = "Percent MT\ntranscripts",
    limits = c(0,1),
    labels = scales::percent,
    guide = guide_colorbar(frame.colour = "black", ticks.colour = "black")
  )

ggsave(paste0(path_results, "plots/qc_nFeature_over_nCount_before_filtering_mt.png"), height = 4, width = 6)

cells %>% 
  ggplot(aes(nCount, nFeature, color = sample)) +
  geom_point(size = 0.1) +
  scale_x_continuous(name = "Number of transcripts", labels = scales::comma) +
  scale_y_continuous(name = "Number of expressed genes", labels = scales::comma) +
  theme_bw() +
  scale_color_manual(name = 'Sample', values = colors$sample) +
  guides(colour = guide_legend(override.aes = list(size=3)))

ggsave(paste0(path_results, "plots/qc_nFeature_over_nCount_before_filtering_samples.png"), height = 4, width = 6)
```

#### Define adaptive thresholds for filtering
```{r, include=TRUE, warning=FALSE, message=FALSE}
cells <- cells 

median_nCount <- median(cells$nCount) 

mad_nCount <- mad(cells$nCount) 

median_nFeature <- median(cells$nFeature) 

mad_nFeature <- mad(cells$nFeature) 

median_percent_MT <- median(cells$percent_MT) 

mad_percent_MT <- mad(cells$percent_MT)
```

```{r, include=TRUE, warning=FALSE, message=FALSE}
median_mads <- as.data.frame(list(
  median_nCount = median_nCount,
  mad_nCount = mad_nCount,
  median_nFeature = median_nFeature,
  mad_nFeature = mad_nFeature,
  median_percent_MT = median_percent_MT,
  mad_percent_MT = mad_percent_MT
)) 

write_csv(median_mads, paste0(path_results, 'tables/median_mads_merged.csv'))

median_mads
```

#### Set thresholds 
It is generally advised to stay within 3-5 MADs around the median.
```{r, warning=FALSE, message=FALSE}

thresholds_nCount <- c(500, median_nCount + 12*mad_nCount)

thresholds_nFeature <- c(250, median_nFeature + 5*mad_nFeature) 

thresholds_percent_MT <- c(0, median_percent_MT + 4*mad_percent_MT)
```

```{r, include=TRUE, warning=FALSE, message=FALSE}
thresholds <- as.data.frame(list(
  thresholds_nCount = thresholds_nCount,
  thresholds_nFeature = thresholds_nFeature,
  thresholds_percent_MT = thresholds_percent_MT
)) 

write_csv(thresholds, paste0(path_results, 'tables/thresholds_merged.csv'))

thresholds
```

#### Plot thresholds 
```{r}
p1 <- ggplot(cells, aes(nCount, fill = sample)) +
  geom_histogram(binwidth = 20) +
  geom_vline(xintercept = median_nCount, color = "cyan") +
  geom_vline(xintercept = thresholds_nCount, color = "red") +
  theme_bw() +
  scale_y_continuous(labels = scales::comma) +
  scale_fill_manual(name = 'Sample', values = colors$sample) +
  labs(title = "Number of transcripts") +
  facet_wrap(~sample, nrow = length(samples)) +
  theme(legend.position = 'none')

p2 <- ggplot(cells, aes(nFeature, fill = sample)) +
  geom_histogram(binwidth = 20) +
  geom_vline(xintercept = median_nFeature, color = "cyan") +
  geom_vline(xintercept = thresholds_nFeature, color = "red") +
  theme_bw() +
  scale_y_continuous(labels = scales::comma) +
  scale_fill_manual(name = 'Sample', values = colors$sample) +
  labs(title = "Number of expressed genes") +
  facet_wrap(~sample, nrow = length(samples)) +
  theme(legend.position = 'none')

p3 <- ggplot(cells, aes(percent_MT, fill = sample)) +
  geom_histogram(binwidth = 0.005) +
  geom_vline(xintercept = median_percent_MT, color = "cyan") +
  geom_vline(xintercept = thresholds_percent_MT, color = "red") +
  theme_bw() +
  scale_fill_manual(name = 'Sample', values = colors$sample) +
  labs(title = "Percent MT transcripts") +
  facet_wrap(~sample, nrow = length(samples)) +
  theme(legend.position = 'none')

ggsave(paste0(path_results, "plots/qc_histogram_nCount_nFeature_percentMT_thresholds.png"),
  p1 + p2 + p3 + plot_layout(ncol = 3), height = 18, width = 12
)
```

```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
p1 <- ggplot(cells, aes(x = sample, y = nCount, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  geom_hline(yintercept = median_nCount, color = "black") +
  geom_hline(yintercept = thresholds_nCount, color = "red") +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = "Number of transcripts", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p2 <- ggplot(cells, aes(x = sample, y = nCount, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  geom_hline(yintercept = median_nCount, color = "black") +
  geom_hline(yintercept = thresholds_nCount, color = "red") +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = "Number of transcripts", subtitle = "log-scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p3 <- ggplot(cells, aes(x = sample, y = nFeature, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  geom_hline(yintercept = median_nFeature, color = "black") +
  geom_hline(yintercept = thresholds_nFeature, color = "red") +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = "Number of expressed genes", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p4 <- ggplot(cells, aes(x = sample, y = nFeature, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  geom_hline(yintercept = median_nFeature, color = "black") +
  geom_hline(yintercept = thresholds_nFeature, color = "red") +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = "Number of expressed genes", subtitle = "log-scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p5 <- ggplot(cells, aes(x = sample, y = percent_MT, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  geom_hline(yintercept = median_percent_MT, color = "black") +
  geom_hline(yintercept = thresholds_percent_MT, color = "red") +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells$sample))) +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "Percent MT transcripts", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()

ggsave(paste0(path_results, "plots/qc_vln_nCount_nFeature_percentMT_thresholds.png"),
  p1 + p3 + p5 +
    p2 + p4 + plot_layout(ncol = 3),
  height = 7, width = 10
)
```

```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
ggplot(cells, aes(nCount, nFeature, color = percent_MT)) +
  geom_point(size = 0.1) +
  geom_hline(yintercept = thresholds_nFeature, color = "red") +
  geom_vline(xintercept = thresholds_nCount, color = "red") +
  scale_x_continuous(name = "Number of transcripts", labels = scales::comma) +
  scale_y_continuous(name = "Number of expressed genes", labels = scales::comma) +
  theme_bw() +
  scale_color_viridis(
    name = "Percent MT\ntranscripts",
    limits = c(0,1),
    labels = scales::percent,
    guide = guide_colorbar(frame.colour = "black", ticks.colour = "black")
  ) + 
  facet_wrap(~sample, ncol = length(samples)*0.5, nrow = 2)

ggsave(paste0(path_results, "plots/qc_nFeature_over_nCount_thresholds.png"), height = 6, width = 10)
```

Another useful diagnostic involves plotting the proportion of mitochondrial counts against some of the other QC metrics. The aim is to confirm that there are no cells with both large total counts and large mitochondrial counts, to ensure that we are not inadvertently removing high-quality cells that happen to be highly metabolically active.
```{r}
p1 <- ggplot(cells, aes(percent_MT, nCount, color = sample)) +
  geom_point(size = 0.1, alpha = 0.7) +
  geom_hline(yintercept = thresholds_nCount, color = "red") +
  geom_vline(xintercept = thresholds_percent_MT, color = "red") +
  scale_y_continuous(name = "Number of transcripts", labels = scales::comma) +
  scale_x_continuous(name = "Percent MT transcripts", limits = c(0,1), labels = scales::percent) +
  theme_bw() +
  scale_color_manual(values = colors$sample) +
  guides(colour = guide_legend(override.aes = list(size=3)))

p2 <- ggplot(cells, aes(percent_MT, nFeature, color = sample)) +
  geom_point(size = 0.1, alpha = 0.7) +
  geom_hline(yintercept = thresholds_nFeature, color = "red") +
  geom_vline(xintercept = thresholds_percent_MT, color = "red") +
  scale_y_continuous(name = "Number of expressed genes", labels = scales::comma) +
  scale_x_continuous(name = "Percent MT transcripts", limits = c(0,1), labels = scales::percent) +
  theme_bw() +
  scale_color_manual(values = colors$sample) +
  guides(colour = guide_legend(override.aes = list(size=3)))

ggsave(paste0(path_results, "plots/counts_over_percentMT_thresholds.png"),
  p1 + p2 + plot_layout(ncol = 2, widths = 6, heights = 5, guides = 'collect'),
  height = 5, width = 12
)
```

#### Filter cells 
```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
cells_filtered <- cells %>%
  dplyr::filter(
    nCount >= thresholds_nCount[1],
    nCount <= thresholds_nCount[2],
    nFeature >= thresholds_nFeature[1],
    nFeature <= thresholds_nFeature[2],
    percent_MT >= thresholds_percent_MT[1],
    percent_MT <= thresholds_percent_MT[2]
  )

DataFrame(cells_filtered)
```

#### Check after filtering out low-quality cells
```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
p1 <- ggplot(cells_filtered, aes(x = sample, y = nCount, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells_filtered$sample))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = "Number of transcripts", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p2 <- ggplot(cells_filtered, aes(x = sample, y = nCount, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells_filtered$sample))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = "Number of transcripts", subtitle = "log-scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p3 <- ggplot(cells_filtered, aes(x = sample, y = nFeature, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells_filtered$sample))) +
  scale_y_continuous(labels = scales::comma) +
  labs(title = "Number of expressed genes", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p4 <- ggplot(cells_filtered, aes(x = sample, y = nFeature, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells_filtered$sample))) +
  scale_y_log10(labels = scales::comma) +
  labs(title = "Number of expressed genes", subtitle = "log-scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()
p5 <- ggplot(cells_filtered, aes(x = sample, y = percent_MT, fill = sample)) +
  geom_violin(draw_quantiles = c(0.5), scale = "area", trim = FALSE) +
  theme_bw() +
  scale_fill_manual(values = colors$sample) +
  scale_x_discrete(limits = rev(levels(cells_filtered$sample))) +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "Percent MT transcripts", subtitle = "linear scale") +
  theme(
    axis.title = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  coord_flip()

ggsave(paste0(path_results, "plots/qc_vln_nCount_nFeature_percentMT_filtered.png"),
  p1 + p3 + p5 +
    p2 + p4 + plot_layout(ncol = 3),
  height = 7, width = 10
)
```

```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
ggplot(cells_filtered, aes(nCount, nFeature, color = percent_MT)) +
  geom_point(size = 0.1) +
  scale_x_continuous(name = "Number of transcripts", labels = scales::comma) +
  scale_y_continuous(name = "Number of expressed genes", labels = scales::comma) +
  theme_bw() +
  scale_color_viridis(
    name = "Percent MT\ntranscripts",
    limits = c(0,1),
    labels = scales::percent,
    guide = guide_colorbar(frame.colour = "black", ticks.colour = "black")
  ) +
  facet_wrap(~sample, ncol = length(samples)*0.5, nrow = 2)

ggsave(paste0(path_results, "plots/qc_nFeature_over_nCount_filtered.png"),  height = 6, width = 10)
```

```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
cells_to_keep <- cells_filtered$cell

length(cells_to_keep) # 45493
```

### remove genes which are expressed in fewer than 5 cells
```{r, warning=FALSE, message=FALSE}
genes <- tibble(
  gene = rownames(transcripts$raw$merged),
  count = Matrix::rowSums(transcripts$raw$merged),
  cells = Matrix::rowSums(transcripts$raw$merged != 0)
)
genes_to_keep <- genes %>% dplyr::filter(cells >= 5) %>% pull(gene)
length(genes_to_keep) # 26614
```

## Generate filtered high-quality matrix
```{r, warning=FALSE, message=FALSE}
transcripts$raw$filtered <- transcripts$raw$merged[genes_to_keep,cells_to_keep]
```

```{r, warning=FALSE, message=FALSE}
cells_qc <- tibble(
  sample = character(),
  n_cells_before = numeric(),
  n_cells_after = numeric()
)
for ( i in samples ) {
  tmp <- tibble(
    sample = i,
    n_cells_before = grep(colnames(transcripts$raw$merged), pattern = paste0("-", i, "$"))
    %>% length(),
    n_cells_after = grep(colnames(transcripts$raw$filtered), pattern = paste0("-", i,
                                                                      "$")) %>% length()
  )
  cells_qc <- bind_rows(cells_qc, tmp)
}

write_csv(cells_qc, paste0(path_results, 'tables/cells_before_after_qc.csv'))
cells_qc
```

```{r}
saveRDS(transcripts$raw$filtered, paste0(path_data, 'merged_matrix_filtered.rds'))
```

## Create Seurat object of AML samples (merged)
```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
merged_seurat <- CreateSeuratObject(
  counts = transcripts$raw$filtered,
  min.cells = 0,
  min.features = 0
)

#store sample assignment for every cell in the meta data
merged_seurat@meta.data$sample <- Cells(merged_seurat) %>%
  strsplit("-") %>%
  vapply(FUN.VALUE = character(1), `[`, 2) %>%
  factor(levels = samples)
```

```{r, warning=FALSE, message=FALSE, out.width="75%", fig.align="center"}
#check again the distribution of number of transcripts and expressed genes per cell by sample 
temp_labels <- merged_seurat@meta.data %>%
  group_by(sample) %>%
  tally()

p1 <- ggplot() +
  geom_half_violin(
    data = merged_seurat@meta.data, aes(sample, nCount_RNA, fill = sample),
    side = "l", show.legend = FALSE, trim = FALSE
  ) +
  geom_half_boxplot(
    data = merged_seurat@meta.data, aes(sample, nCount_RNA, fill = sample),
    side = "r", outlier.color = NA, width = 0.4, show.legend = FALSE
  ) +
  geom_text(
    data = temp_labels,
    aes(x = sample, y = -Inf, label = paste0("n = ", format(n, big.mark = ",",
                                                            trim = TRUE)), vjust = -1),
    color = "black", size = 2.8
  ) +
  scale_color_manual(values = colors$sample) +
  scale_fill_manual(values = colors$sample) +
  scale_y_continuous(labels = scales::comma, expand = c(0.08,0)) +
  theme_bw() +
  labs(x = "", y = "Number of transcripts") +
  theme(
    panel.grid.major.x = element_blank(),
    axis.title.x = element_blank()
  )

p2 <- ggplot() +
  geom_half_violin(
    data = merged_seurat@meta.data, aes(sample, nFeature_RNA, fill = sample),
    side = "l", show.legend = FALSE, trim = FALSE
  ) +
  geom_half_boxplot(
    data = merged_seurat@meta.data, aes(sample, nFeature_RNA, fill = sample),
    side = "r", outlier.color = NA, width = 0.4, show.legend = FALSE
  ) +
  geom_text(
    data = temp_labels,
    aes(x = sample, y = -Inf, label = paste0("n = ", format(n, big.mark = ",",
                                                            trim = TRUE)), vjust = -1),
    color = "black", size = 2.8
  ) +
  scale_color_manual(values = colors$sample) +
  scale_fill_manual(values = colors$sample) +
  scale_y_continuous(
    name = "Number of expressed genes",
    labels = scales::comma, expand = c(0.08,0)
  ) +
  theme_bw() +
  theme(
    panel.grid.major.x = element_blank(),
    axis.title.x = element_blank()
  )

ggsave(paste0(path_results, "plots/nCount_nFeature_by_sample.png"),
  p1 + p2 + plot_layout(ncol = 2), height = 4, width = 15
)
```

## Add percentage of mitochondrial genes in metadata
```{r}
DefaultAssay(object = merged_seurat) <- "RNA"
merged_seurat[["percent_MT"]] <- PercentageFeatureSet(merged_seurat, "MT-")
```

## Log normalization 
Seurat log-normalization is needed for mapping cell lineage to Triana's reference
```{r}
merged_seurat <- NormalizeData(object = merged_seurat, 
                               normalization.method = "LogNormalize",
                               scale.factor = 10000)
merged_seurat <- FindVariableFeatures(object = merged_seurat, selection.method = "vst") # 2000 by default
```

## Cell cycle analysis
```{r, warning=FALSE, message=FALSE}
merged_seurat <- CellCycleScoring(
  merged_seurat,
  s.features = cc.genes.updated.2019$s.genes,
  g2m.features = cc.genes.updated.2019$g2m.genes,
  assay = 'RNA'
)
merged_seurat@meta.data$cell_cycle_seurat <- merged_seurat@meta.data$Phase
merged_seurat@meta.data$Phase <- NULL
merged_seurat@meta.data$cell_cycle_seurat <- factor(
  merged_seurat@meta.data$cell_cycle_seurat, levels = c("G1", "S", "G2M")
)
```

## Set cohort metadata
```{r}
merged_seurat$cohort <- ifelse(
  merged_seurat$sample %in% c('AML4', 'AML5', 'AML6', 'sAML1'), 'SRSF2mut', 
  ifelse(merged_seurat$sample %in% c('AML2', 'AML3', 'sAML16'), 'SRSF2wt', 'hBM'
  )
)

merged_seurat$cohort <- factor(merged_seurat$cohort, levels = c("SRSF2mut", "SRSF2wt", 'hBM'))
```

## Set sequencing run metadata
```{r}
merged_seurat$seq_run <- case_when(
  merged_seurat$sample %in% c("AML4", "AML5") ~ "batch1",
  merged_seurat$sample == "AML6" ~ "batch2",
  merged_seurat$sample == "sAML1" ~ "batch3",
  merged_seurat$sample %in% c("AML2", "sAML16") ~ "batch4",
  merged_seurat$sample == "AML3" ~ "batch5",
  merged_seurat$sample %in% c("hBM1", "hBM2", "hBM3") ~ "batch6"
  )

merged_seurat$seq_run <- factor(merged_seurat$seq_run, levels = paste0("batch", 1:6))
```

## Save log-normalized merged Seurat object 
```{r}
saveRDS(merged_seurat, paste0(path_data, 'merged_seurat_logcounts.rds'))
```

## Session info
```{r}
sessionInfo()
```
