---
title: "motif"
author: "Iman"
date: "2023-07-25"
output: html_document
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
defaultW <- getOption("warn")
options(device = "RStudioGD")
```

## Loading libraries

```{r include=FALSE, results='hide'}
library(devtools)
library(stringr)
library(GenomicFeatures)
library(AnnotationDbi)
library(Biostrings)
library(AnnotationHub)
library(ensembldb)
library(tidyverse)
library(dplyr)
library(DECIPHER)
library(ggplot2)
library(enrichR)
library(ComplexHeatmap)
library(circlize)
library(DOSE)
library(org.Hs.eg.db)
library(AnnotationDbi)
library('biomaRt')
library(enrichplot)
library(clusterProfiler)
library(GOSemSim)
library(annotables)
library(gprofiler2)
```

## extracting reads inprimer exon and adjecent ones
```{r}

pattern <- "chr(\\d+):(\\d+)-(\\d+)_TCONS\\.(\\d+)_SES_(Up|Down)Regulated"
events_splicing<-readRDS("/Users/ieo5268/Documents/Thesis_Iman/cluster/reuslts/splicing/results/splicing/events_object.RDS")
gtf_file <- "/Users/ieo5268/Documents/Thesis_Iman/cluster/GTF/gffcomare_AML_12345_hBM_123.combined.gtf"
df_split <- read.delim(gtf_file, header = F, stringsAsFactors = FALSE)
####
gtf_data <- df_split %>%
  separate(V9, into = c("transcript_id", "gene_id", "exon_number"), sep = ";",
           extra = "drop", remove = FALSE) %>%
  extract(col = transcript_id, into = "transcript_id", regex = "transcript_id\\s([^;]+)", remove = FALSE) %>%
  extract(col = exon_number, into = "exon_number", regex = "exon_number\\s([^;]+)", remove = FALSE)

######
region_within_exon <- function(region_start, region_end, exon_start, exon_end) {
  return(region_start >= exon_start & region_end <= exon_end)
}

#####
extract_info <- function(string) {
  result <- str_match(string, pattern)
  if (!is.na(result[1, 2])) {
    chromosome <- paste0("chr",result[1, 2])
    start <- as.numeric(result[1, 3])
    end <- as.numeric(result[1, 4])
    Transcript <- paste0("TCONS_",result[1, 5])
    regulation <- result[1, 6]
    
    return(data.frame(chromosome = chromosome, start = start, end = end,Transcript=Transcript, regulation = regulation))
  }
  return(NULL)
}
#####
samples<- c("sAML1","AML4","AML5")
bam_hM <- paste0("/Volumes/scratch/PGP/SCMseq/AML_3wtSample/hBM1/FLAMES_out/align2genome.bam")
results_motif<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/hBM1/"
dir.create(results_motif)
###### extrating region of Healthy BM
for(i in samples){
ses<- events_splicing[[i]]$SES_filtered$gene_evnt_ex
events_splicing[[i]]$MES_filtered$gene_evnt_ex

# Extract the information and create a data frame
output_df <- do.call(rbind, lapply(ses, extract_info))
output_df_mut<- output_df%>% dplyr::filter(regulation=="Up")

mut_ls<-sapply(c(1:length(output_df_mut$chromosome)),function(r){
overlapping_exons <- gtf_data %>%
  filter(gtf_data$V1 == output_df_mut$chromosome[r] & gtf_data$V3 == "exon" &
         region_within_exon(as.integer(output_df_mut$start[r]), as.integer(output_df_mut$end[r]), gtf_data$V4, gtf_data$V5) &  gtf_data$transcript_id == output_df_mut$Transcript[r])
exon_id=as.integer(overlapping_exons$exon_number)
### finding start of the previous exon and end of the next exon
start_df<- gtf_data %>% filter(gtf_data$V1 == overlapping_exons$V1 & gtf_data$V3 == "exon" & gtf_data$transcript_id==overlapping_exons$transcript_id & gtf_data$exon_number==exon_id-1)
end_df<- gtf_data %>% filter(gtf_data$V1 == overlapping_exons$V1 & gtf_data$V3 == "exon" & gtf_data$transcript_id==overlapping_exons$transcript_id & gtf_data$exon_number==exon_id+1)

gr_ex0<- GRanges(seqnames =overlapping_exons$V1 , IRanges(start_df$V5-49, start_df$V5))
gr_ex1_0<- GRanges(seqnames =overlapping_exons$V1 , IRanges(overlapping_exons$V4, overlapping_exons$V4+49))
gr_ex1_1<- GRanges(seqnames =overlapping_exons$V1 , IRanges(overlapping_exons$V5-49, overlapping_exons$V5))
gr_ex2<- GRanges(seqnames =overlapping_exons$V1 , IRanges(end_df$V4, end_df$V4+49))

#which <- GRanges(seqnames =overlapping_exons$V1 , IRanges(start_df$V4, end_df$V5))
p0 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex0)
p1_0 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex1_0)
p1_1 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex1_1)
p2 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex2)


filtered_bam_mut_0<-GenomicAlignments::stackStringsFromBam(bam_hM,param = p0,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_mut_1_0<-GenomicAlignments::stackStringsFromBam(bam_hM,param = p1_0,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_mut_1_1<-GenomicAlignments::stackStringsFromBam(bam_hM,param = p1_1,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_mut_2<-GenomicAlignments::stackStringsFromBam(bam_hM,param = p2,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")

x_0<- Biostrings::xscat(filtered_bam_mut_0,filtered_bam_mut_1_0)
x_1<- Biostrings::xscat(filtered_bam_mut_1_1,filtered_bam_mut_2)
#names(x)<-seq(length(x))
return(list(x_0,x_1))
})


combined_vector_0 <- do.call(c, mut_ls[1,])
names(combined_vector_0)<-seq(length(combined_vector_0))
writeXStringSet(combined_vector_0,filepath = paste0(results_motif,i,"_mut_SES_hbm_0.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_0,htmlFile =paste0(results_motif,i,"_mut_SES_hbm_0.html") )

combined_vector_1 <- do.call(c, mut_ls[2,])
names(combined_vector_1)<-seq(length(combined_vector_1))
writeXStringSet(combined_vector_1,filepath = paste0(results_motif,i,"_mut_SES_hbm_1.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_1,htmlFile =paste0(results_motif,i,"_mut_SES_hbm_1.html") )

}

########################
results_motif<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/DNA_binding_sites/"
####### region of mut and wt
for(i in samples){
  i="AML4"
ses<- events_splicing[[i]]$SES_filtered$gene_evnt_ex

# Extract the information and create a data frame
output_df <- do.call(rbind, lapply(ses, extract_info))
output_df_mut<- output_df%>% dplyr::filter(regulation=="Up")
output_df_wt<- output_df%>% dplyr::filter(regulation=="Down")

bam_mut <- paste0("/Users/ieo5268/Documents/Thesis_Iman/cluster/bam_AML4/",i,"_mut_sorted.bam")
bam_wt <- paste0("/Users/ieo5268/Documents/Thesis_Iman/cluster/bam_AML4/",i,"_wt_sorted.bam")


mut_ls<-sapply(c(1:length(output_df_mut$chromosome)),function(r){
  r=1
overlapping_exons <- gtf_data %>%
  filter(gtf_data$V1 == output_df_mut$chromosome[r] & gtf_data$V3 == "exon" &
         region_within_exon(as.integer(output_df_mut$start[r]), as.integer(output_df_mut$end[r]), gtf_data$V4, gtf_data$V5) &  gtf_data$transcript_id == output_df_mut$Transcript[r])
exon_id=as.integer(overlapping_exons$exon_number)
### finding start of the previous exon and end of the next exon
start_df<- gtf_data %>% filter(gtf_data$V1 == overlapping_exons$V1 & gtf_data$V3 == "exon" & gtf_data$transcript_id==overlapping_exons$transcript_id & gtf_data$exon_number==exon_id-1)
end_df<- gtf_data %>% filter(gtf_data$V1 == overlapping_exons$V1 & gtf_data$V3 == "exon" & gtf_data$transcript_id==overlapping_exons$transcript_id & gtf_data$exon_number==exon_id+1)

gr_ex0<- GRanges(seqnames =overlapping_exons$V1 , IRanges(start_df$V5-49, start_df$V5))
gr_ex1_0<- GRanges(seqnames =overlapping_exons$V1 , IRanges(overlapping_exons$V4, overlapping_exons$V4+49))
gr_ex1_1<- GRanges(seqnames =overlapping_exons$V1 , IRanges(overlapping_exons$V5-49, overlapping_exons$V5))
gr_ex2<- GRanges(seqnames =overlapping_exons$V1 , IRanges(end_df$V4, end_df$V4+49))


#which <- GRanges(seqnames =overlapping_exons$V1 , IRanges(start_df$V4, end_df$V5))
p0 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex0)
p1_0 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex1_0)
p1_1 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex1_1)
p2 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex2)


filtered_bam_mut_0<-GenomicAlignments::stackStringsFromBam(bam_mut,param = p0,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
GenomicAlignments::stackStringsFromBam(bam_mut,param = p0,use.names = T,what = "qual") %>% as.data.frame()
filtered_bam_mut_1_0<-GenomicAlignments::stackStringsFromBam(bam_mut,param = p1_0,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_mut_1_1<-GenomicAlignments::stackStringsFromBam(bam_mut,param = p1_1,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_mut_2<-GenomicAlignments::stackStringsFromBam(bam_mut,param = p2,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")

x_0<- Biostrings::xscat(filtered_bam_mut_0,filtered_bam_mut_1_0) ####  double check
x_1<- Biostrings::xscat(filtered_bam_mut_1_1,filtered_bam_mut_2)

#names(x)<-seq(length(x))
return(list(x_0,x_1))
})


combined_vector_0 <- do.call(c, mut_ls[1,])
names(combined_vector_0)<-seq(length(combined_vector_0))
writeXStringSet(combined_vector_0,filepath = paste0(results_motif,i,"_SES_mut_0.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_0,htmlFile =paste0(results_motif,i,"_SES_mut_0.html") )

combined_vector_1 <- do.call(c, mut_ls[2,])
names(combined_vector_1)<-seq(length(combined_vector_1))
writeXStringSet(combined_vector_1,filepath = paste0(results_motif,i,"_SES_mut_1.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_1,htmlFile =paste0(results_motif,i,"_SES_mut_1.html") )
######################## wt

wt_ls<-sapply(c(1:length(output_df_wt$chromosome)),function(r){
overlapping_exons <- gtf_data %>%
  filter(gtf_data$V1 == output_df_wt$chromosome[r] & gtf_data$V3 == "exon" &
         region_within_exon(as.integer(output_df_wt$start[r]), as.integer(output_df_wt$end[r]), gtf_data$V4, gtf_data$V5) &  gtf_data$transcript_id == output_df_wt$Transcript[r])
exon_id=as.integer(overlapping_exons$exon_number)
### finding start of the previous exon and end of the next exon
start_df<- gtf_data %>% filter(gtf_data$V1 == overlapping_exons$V1 & gtf_data$V3 == "exon" & gtf_data$transcript_id==overlapping_exons$transcript_id & gtf_data$exon_number==exon_id-1)
end_df<- gtf_data %>% filter(gtf_data$V1 == overlapping_exons$V1 & gtf_data$V3 == "exon" & gtf_data$transcript_id==overlapping_exons$transcript_id & gtf_data$exon_number==exon_id+1)

gr_ex0<- GRanges(seqnames =overlapping_exons$V1 , IRanges(start_df$V5-49, start_df$V5))
gr_ex1_0<- GRanges(seqnames =overlapping_exons$V1 , IRanges(overlapping_exons$V4, overlapping_exons$V4+49))
gr_ex1_1<- GRanges(seqnames =overlapping_exons$V1 , IRanges(overlapping_exons$V5-49, overlapping_exons$V5))
gr_ex2<- GRanges(seqnames =overlapping_exons$V1 , IRanges(end_df$V4, end_df$V4+49))

#which <- GRanges(seqnames =overlapping_exons$V1 , IRanges(start_df$V4, end_df$V5))
p0 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex0)
p1_0 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex1_0)
p1_1 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex1_1)
p2 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex2)


filtered_bam_wt_0<-GenomicAlignments::stackStringsFromBam(bam_wt,param = p0,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_wt_1_0<-GenomicAlignments::stackStringsFromBam(bam_wt,param = p1_0,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_wt_1_1<-GenomicAlignments::stackStringsFromBam(bam_wt,param = p1_1,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")
filtered_bam_wt_2<-GenomicAlignments::stackStringsFromBam(bam_wt,param = p2,use.names = T,what = "seq",Lpadding.letter = "N",Rpadding.letter = "N",D.letter = "N",N.letter = "N")

x_0<- Biostrings::xscat(filtered_bam_wt_0,filtered_bam_wt_1_0)
x_1<- Biostrings::xscat(filtered_bam_wt_1_1,filtered_bam_wt_2)

return(list(x_0,x_1))
})

combined_vector_0 <- do.call(c, wt_ls[1,])
names(combined_vector_0)<-seq(length(combined_vector_0))
writeXStringSet(combined_vector_0,filepath = paste0(results_motif,i,"_SES_wt_0.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_0,htmlFile =paste0(results_motif,i,"_SES_wt_0.html") )

combined_vector_1 <- do.call(c, wt_ls[2,])
names(combined_vector_1)<-seq(length(combined_vector_1))
writeXStringSet(combined_vector_1,filepath = paste0(results_motif,i,"_SES_wt_1.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_1,htmlFile =paste0(results_motif,i,"_SES_wt_1.html") )
}

```

## extractin reads just for primer exon

```{r}

pattern <- "chr(\\d+):(\\d+)-(\\d+)_TCONS\\.(\\d+)_SES_(Up|Down)Regulated"
events_splicing<-readRDS("/Users/ieo5268/Documents/Thesis_Iman/cluster/reuslts/splicing/results/splicing/events_object.RDS")
gtf_file <- "/Users/ieo5268/Documents/Thesis_Iman/cluster/GTF/gffcomare_AML_12345_hBM_123.combined.gtf"
df_split <- read.delim(gtf_file, header = F, stringsAsFactors = FALSE)
####
gtf_data <- df_split %>%
  separate(V9, into = c("transcript_id", "gene_id", "exon_number"), sep = ";",
           extra = "drop", remove = FALSE) %>%
  extract(col = transcript_id, into = "transcript_id", regex = "transcript_id\\s([^;]+)", remove = FALSE) %>%
  extract(col = exon_number, into = "exon_number", regex = "exon_number\\s([^;]+)", remove = FALSE)

######
region_within_exon <- function(region_start, region_end, exon_start, exon_end) {
  return(region_start >= exon_start & region_end <= exon_end)
}

#####
extract_info <- function(string) {
  result <- str_match(string, pattern)
  if (!is.na(result[1, 2])) {
    chromosome <- paste0("chr",result[1, 2])
    start <- as.numeric(result[1, 3])
    end <- as.numeric(result[1, 4])
    Transcript <- paste0("TCONS_",result[1, 5])
    regulation <- result[1, 6]
    
    return(data.frame(chromosome = chromosome, start = start, end = end,Transcript=Transcript, regulation = regulation))
  }
  return(NULL)
}
#####
samples<- c("sAML1","AML4","AML5")
results_motif<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/NUCB2/2nd_try/"
dir.create(results_motif)
###### extrating region of Healthy BM
for(i in samples){
  
ses<- events_splicing[[i]]$SES_filtered$gene_evnt_ex
#events_splicing[[i]]$MES_filtered$gene_evnt_ex

# Extract the information and create a data frame
output_df <- do.call(rbind, lapply(ses, extract_info))
output_df_mut<- output_df%>% dplyr::filter(regulation=="Up")
bam_m<- paste0("/Volumes/scratch/PGP/SCMseq/isoform_analysis/fastq_split/fastq_spli_allTR/Realignment_bam_by_genotype_merged_gtf/",i,"_mut_sorted.bam")
bam_w<- paste0("/Volumes/scratch/PGP/SCMseq/isoform_analysis/fastq_split/fastq_spli_allTR/Realignment_bam_by_genotype_merged_gtf/",i,"_wt_sorted.bam")
mut_ls<-sapply(c(1:length(output_df_mut$chromosome)),function(r){
overlapping_exons <- gtf_data %>%
  filter(gtf_data$V1 == output_df_mut$chromosome[r] & gtf_data$V3 == "exon" &
         region_within_exon(as.integer(output_df_mut$start[r]), as.integer(output_df_mut$end[r]), gtf_data$V4, gtf_data$V5) &  gtf_data$transcript_id == output_df_mut$Transcript[r])
exon_id=as.integer(overlapping_exons$exon_number)
### finding start of the previous exon and end of the next exon
gr_ex1_0<- GRanges(seqnames =overlapping_exons$V1 , IRanges(overlapping_exons$V4, overlapping_exons$V4+49))
p1_0 <- Rsamtools::ScanBamParam(mapqFilter = 0,which = gr_ex1_0)



filtered_bam_mut_1_0_s<-GenomicAlignments::stackStringsFromBam(bam_m,param = p1_0,use.names = T,what = "seq",D.letter = "N",N.letter = "N")
filtered_bam_wt_1_0_s<-GenomicAlignments::stackStringsFromBam(bam_w,param = p1_0,use.names = T,what = "seq",D.letter = "N",N.letter = "N")
Ncount <- alphabetFrequency(DNAStringSet(filtered_bam_mut_1_0_s))# Generates N counts for each sequence in dvec.
filtered_bam_mut_1_0_s<- filtered_bam_mut_1_0_s[(Ncount[,"N"] <= 5) & (Ncount[,"+"]==0)]### selectin reads with 10% N just

Ncount <- alphabetFrequency(DNAStringSet(filtered_bam_wt_1_0_s))# Generates N counts for each sequence in dvec.
filtered_bam_wt_1_0_s<- filtered_bam_wt_1_0_s[(Ncount[,"N"] <= 5) & (Ncount[,"+"]==0)]### selectin reads with 10% N just


return(list(filtered_bam_mut_1_0_s,filtered_bam_wt_1_0_s))
})

combined_vector_0 <- do.call(c, mut_ls[1,])
names(combined_vector_0)<-seq(length(combined_vector_0))
writeXStringSet(combined_vector_0,filepath = paste0(results_motif,i,"_mut_SES.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_0,htmlFile =paste0(results_motif,i,"_mut_SES.html") )

combined_vector_1 <- do.call(c, mut_ls[2,])
names(combined_vector_1)<-seq(length(combined_vector_1))
writeXStringSet(combined_vector_1,filepath = paste0(results_motif,i,"_wt_SES.fa"),compress = F,append = F,format = "fasta")
DECIPHER::BrowseSeqs(combined_vector_1,htmlFile =paste0(results_motif,i,"_wt_SES.html") )

}
```
```{r}
# running TOMTOM on the memesuit website to find genes associated to those motifs

```

## Importing data from TOMTOM output for DNA binding sites
```{r}

results_motif<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/DNA_binding_sites/"
dir.create(results_motif)
hsGO <- godata('org.Hs.eg.db', ont="MF")
ensembl=useMart("ensembl")
ensembl_d = useDataset("hsapiens_gene_ensembl",mart=ensembl)
tt<-lapply(samples,function(i){
  t<- lapply(c(0,1),function(p){
  x<- read.csv(paste0("/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/DNA_binding_sites/meme_out_",i,"_SES_",p,"/tomtom.tsv"),sep = "\t")
  #x<- read.csv(paste0("/Users/ieo5268/Documents/Thesis_Iman/tomtom (2).tsv"),sep = "\t")
  #summary(x$p.value)
  #x<- separate(x,Target_ID,into = c("gene","database"),sep = "_")
  
  x<- separate(x,Target_ID,into = c("gene","database"),sep = "_")
  x<- x[complete.cases(x),]$gene
  #x<-motif_data
  ##############
  gostres <- gost(query = x, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)
  
  p1<- gostplot(gostres, capped = T, interactive = F)
  p1<- p1+theme(text = element_text(size = 10))
  p1
  ggsave(paste0(results_motif,i,"_",p,"_gosplot_.","jpg"),plot = p1,width = 15,height = 7)
  publish_gosttable(gostres, highlight_terms = gostres$result[gostres$result$highlighted==T,]$term_id,
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size", "intersection_size"),
                        filename = paste0(results_motif,i,"_",p,"_gostable_.","pdf"))
  ############
  
  G_list <- getBM(attributes = c('entrezgene_id','hgnc_symbol',"ensembl_gene_id"), filters = 'hgnc_symbol', values =x, mart = ensembl_d)
  enr_bp<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "BP",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_cc<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "CC",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_mf<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "MF",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  a1<- barplot(enr_cc)+ labs(subtitle = paste0("Enriched_GO_","CC"))
  a2<- barplot(enr_bp)+ labs(subtitle = paste0("Enriched_GO_","BP"))
  a3<- barplot(enr_mf)+ labs(subtitle = paste0("Enriched_GO_","MF"))

  q<-cowplot::plot_grid( a1,a2,a3, ncol=3)
  ggsave(paste0(results_motif,i,"_",p,"_enriched_GO_123_",".jpg"),plot = q,width = 30,height = 7)
  
  kk <- enrichKEGG(gene = G_list$entrezgene_id,organism   = 'hsa', pvalueCutoff = 0.05)
  edox <- setReadable(kk, 'org.Hs.eg.db', 'ENTREZID')
  a1<- dotplot(edox)
  a2<- cnetplot(edox, categorySize="geneNum")
  q<-cowplot::plot_grid( a1,a2, ncol=2)
  ggsave(paste0(results_motif,i,"_",p,"_KEGG_",".jpg"),plot = q,width = 18,height = 7)
  
  pdf(paste0(results_motif,i,"_",p,"_enrichedGO_BP_",".pdf"),width = 12,height = 12)
  plotGOgraph(enr_bp)
  dev.off()
  
  
return(G_list)
  })
  names(t)<- c("fistr_half","seconed_half")
  return(t)
}
)
names(tt)<- samples
saveRDS(object = tt,file = paste0(results_motif,"tomtom_object_DNA.rds"))

ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
  list1=list()
  list1$sAML1_0<- tt$sAML1$fistr_half$hgnc_symbol
  list1$AML4_0<- tt$AML4$fistr_half$hgnc_symbol
  list1$AML5_0<- tt$AML5$fistr_half$hgnc_symbol
com_gene_0<- Reduce(intersect,list1)%>%as.data.frame()
geneDescriptions_0 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_0, mart = ensembl)

####
list2=list()
  list2$sAML1_1<- tt$sAML1$seconed_half$hgnc_symbol
  list2$AML4_1<- tt$AML4$seconed_half$hgnc_symbol
  list2$AML5_1<- tt$AML5$seconed_half$hgnc_symbol
com_gene_1<- Reduce(intersect,list2)
geneDescriptions_1 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_1, mart = ensembl)

  ####

com_1_0<-intersect(com_gene_0,com_gene_1)

desc<- grch38 %>% filter(symbol %in% com_gene_1) %>% dplyr::select(ensgene, symbol, chr, start, end, description)

# edo <- DOSE::enrichDGN(G_list$entrezgene_id)
#   q<-dotplot(edo, showCategory=20) + labs(title = "Dotplot of the enriched pathways in NCG",subtitle =paste0(i,"_",p,"_NCG"))+
#   theme_bw() +theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 10),axis.text.y = element_text( size = 12))
# plot(q)
# ggsave(paste0(results_motif,i,"_",p,"_NCG",".jpg"),plot = q,width = 15,height = 15)
# edox <- setReadable(edo, 'org.Hs.eg.db', 'ENTREZID')
# p1 <- cnetplot(edox, foldChange=x)+  labs(subtitle = "Network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# p2 <- cnetplot(edox, foldChange=x, circular = TRUE, colorEdge = TRUE) +labs(subtitle = "Circular network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# q<-cowplot::plot_grid(p1, p2, ncol=2, labels=LETTERS[1:2],label_size = 1)
# ggsave(paste0(results_motif,i,"_",p,"_NCG_network",".jpg"),plot = q,width = 20,height = 10)
# 
# p1 <- heatplot(edox, showCategory=20)
# p2 <- heatplot(edox, foldChange=edo, showCategory=20)
# q<-cowplot::plot_grid(p1, p2, ncol=1, labels=LETTERS[1:2])
# ggsave(paste0(results_motif,i,"_",p,"_NCG_heatmap",".jpg"),plot = q,width = 20,height = 15)
# 


```




## Importing data from TOMTOM output for RNA binding sites
```{r}
results_motif<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/RNA_binding_sites/"

ensembl=useMart("ensembl")
ensembl_d = useDataset("hsapiens_gene_ensembl",mart=ensembl)
tt<-lapply(samples,function(i){
  t<- lapply(c(0,1),function(p){
 xml_file_path <- paste0("/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/RNA_binding_sites/",i,"_tomtom_",p,".xml")
#xml_file_path<-"/Users/ieo5268/Documents/Thesis_Iman/tomtom.xml"
# Parse the XML file
xml_data <- xmlParse(xml_file_path)

# Access elements and attributes in the XML file
#x <- getNodeSet(xml_data, "//motif")
motif_elements <- getNodeSet(xml_data, "//motif")
x <- sapply(motif_elements, function(motif) {
xmlGetAttr(motif, "alt")
})
  #x<-motif_data
  ##############
  gostres <- gost(query = x, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)
  
  p1<- gostplot(gostres, capped = T, interactive = F)
  p1<- p1+theme(text = element_text(size = 10))
  p1
  ggsave(paste0(results_motif,i,"_",p,"_gosplot_.","jpg"),plot = p1,width = 15,height = 7)
  publish_gosttable(gostres, highlight_terms = gostres$result[gostres$result$highlighted==T,]$term_id,
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size", "intersection_size"),
                        filename = paste0(results_motif,i,"_",p,"_gostable_.","pdf"))
  ############
  
  G_list <- getBM(attributes = c('entrezgene_id','hgnc_symbol',"ensembl_gene_id"), filters = 'hgnc_symbol', values =x, mart = ensembl_d)
  enr_bp<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "BP",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_cc<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "CC",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_mf<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "MF",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  a1<- barplot(enr_cc)+ labs(subtitle = paste0("Enriched_GO_","CC"))
  a2<- barplot(enr_bp)+ labs(subtitle = paste0("Enriched_GO_","BP"))
  a3<- barplot(enr_mf)+ labs(subtitle = paste0("Enriched_GO_","MF"))

  q<-cowplot::plot_grid( a1,a2,a3, ncol=3)
  ggsave(paste0(results_motif,i,"_",p,"_enriched_GO_123_",".jpg"),plot = q,width = 30,height = 7)
  
  kk <- enrichKEGG(gene = G_list$entrezgene_id,organism   = 'hsa', pvalueCutoff = 0.05)
  edox <- setReadable(kk, 'org.Hs.eg.db', 'ENTREZID')
  a1<- dotplot(edox)
  a2<- cnetplot(edox, categorySize="geneNum")
  q<-cowplot::plot_grid( a1,a2, ncol=2)
  ggsave(paste0(results_motif,i,"_",p,"_KEGG_",".jpg"),plot = q,width = 18,height = 7)
  
  pdf(paste0(results_motif,i,"_",p,"_enrichedGO_BP_",".pdf"),width = 12,height = 12)
  plotGOgraph(enr_bp)
  dev.off()
  
  
return(G_list)
  })
  names(t)<- c("fistr_half","seconed_half")
  return(t)
}
)
names(tt)<- samples
saveRDS(object = tt,file = paste0(results_motif,"tomtom_object_RNA.rds"))

ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
  list1=list()
  list1$sAML1_0<- tt$sAML1$fistr_half$hgnc_symbol
  list1$AML4_0<- tt$AML4$fistr_half$hgnc_symbol
  list1$AML5_0<- tt$AML5$fistr_half$hgnc_symbol
com_gene_0<- Reduce(intersect,list1)%>%as.data.frame()
geneDescriptions_0 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_0, mart = ensembl)

####
list2=list()
  list2$sAML1_1<- tt$sAML1$seconed_half$hgnc_symbol
  list2$AML4_1<- tt$AML4$seconed_half$hgnc_symbol
  list2$AML5_1<- tt$AML5$seconed_half$hgnc_symbol
com_gene_1<- Reduce(intersect,list2)
geneDescriptions_1 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_1, mart = ensembl)

  ####

com_1_0<-intersect(geneDescriptions_1$hgnc_symbol,geneDescriptions_0$hgnc_symbol)

desc<- grch38 %>% filter(symbol %in% com_gene_1) %>% dplyr::select(ensgene, symbol, chr, start, end, description)

# edo <- DOSE::enrichDGN(G_list$entrezgene_id)
#   q<-dotplot(edo, showCategory=20) + labs(title = "Dotplot of the enriched pathways in NCG",subtitle =paste0(i,"_",p,"_NCG"))+
#   theme_bw() +theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 10),axis.text.y = element_text( size = 12))
# plot(q)
# ggsave(paste0(results_motif,i,"_",p,"_NCG",".jpg"),plot = q,width = 15,height = 15)
# edox <- setReadable(edo, 'org.Hs.eg.db', 'ENTREZID')
# p1 <- cnetplot(edox, foldChange=x)+  labs(subtitle = "Network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# p2 <- cnetplot(edox, foldChange=x, circular = TRUE, colorEdge = TRUE) +labs(subtitle = "Circular network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# q<-cowplot::plot_grid(p1, p2, ncol=2, labels=LETTERS[1:2],label_size = 1)
# ggsave(paste0(results_motif,i,"_",p,"_NCG_network",".jpg"),plot = q,width = 20,height = 10)
# 
# p1 <- heatplot(edox, showCategory=20)
# p2 <- heatplot(edox, foldChange=edo, showCategory=20)
# q<-cowplot::plot_grid(p1, p2, ncol=1, labels=LETTERS[1:2])
# ggsave(paste0(results_motif,i,"_",p,"_NCG_heatmap",".jpg"),plot = q,width = 20,height = 15)
# 


```
## running pathway analysis for hBM1
```{r}
results_motif<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/hBM1/results_DNA/"
dir.create(results_motif)
hsGO <- godata('org.Hs.eg.db', ont="MF")
ensembl=useMart("ensembl")
ensembl_d = useDataset("hsapiens_gene_ensembl",mart=ensembl)
tt<-lapply(samples,function(i){
  t<- lapply(c(0,1),function(p){
  x<- read.csv(paste0("/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/hBM1/Tomom_RNA_motif_enrichment_comparison/",i,"_",p,".tsv"),sep = "\t")
  #x<- read.csv(paste0("/Users/ieo5268/Documents/Thesis_Iman/tomtom (2).tsv"),sep = "\t")
  #summary(x$p.value)
  #x<- separate(x,Target_ID,into = c("gene","database"),sep = "_")
  
  x<- separate(x,Target_ID,into = c("gene","database"),sep = "_")
  x<- x[complete.cases(x),]$gene
  #x<-motif_data
  ##############
  gostres <- gost(query = x, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)
  
  p1<- gostplot(gostres, capped = T, interactive = F)
  p1<- p1+theme(text = element_text(size = 10))
  p1
  ggsave(paste0(results_motif,i,"_",p,"_gosplot_.","jpg"),plot = p1,width = 15,height = 7)
  publish_gosttable(gostres, highlight_terms = gostres$result[gostres$result$highlighted==T,]$term_id,
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size", "intersection_size"),
                        filename = paste0(results_motif,i,"_",p,"_gostable_.","pdf"))
  ############
  
  G_list <- getBM(attributes = c('entrezgene_id','hgnc_symbol',"ensembl_gene_id"), filters = 'hgnc_symbol', values =x, mart = ensembl_d)
  enr_bp<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "BP",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_cc<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "CC",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_mf<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "MF",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  a1<- barplot(enr_cc)+ labs(subtitle = paste0("Enriched_GO_","CC"))
  a2<- barplot(enr_bp)+ labs(subtitle = paste0("Enriched_GO_","BP"))
  a3<- barplot(enr_mf)+ labs(subtitle = paste0("Enriched_GO_","MF"))

  q<-cowplot::plot_grid( a1,a2,a3, ncol=3)
  ggsave(paste0(results_motif,i,"_",p,"_enriched_GO_123_",".jpg"),plot = q,width = 30,height = 7)
  
  kk <- enrichKEGG(gene = G_list$entrezgene_id,organism   = 'hsa', pvalueCutoff = 0.05)
  edox <- setReadable(kk, 'org.Hs.eg.db', 'ENTREZID')
  a1<- dotplot(edox)
  a2<- cnetplot(edox, categorySize="geneNum")
  q<-cowplot::plot_grid( a1,a2, ncol=2)
  ggsave(paste0(results_motif,i,"_",p,"_KEGG_",".jpg"),plot = q,width = 18,height = 7)
  
  pdf(paste0(results_motif,i,"_",p,"_enrichedGO_BP_",".pdf"),width = 12,height = 12)
  plotGOgraph(enr_bp)
  dev.off()
  
  
return(G_list)
  })
  names(t)<- c("fistr_half","seconed_half")
  return(t)
}
)
names(tt)<- samples
saveRDS(object = tt,file = paste0(results_motif,"tomtom_object_DNA.rds"))

ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
  list1=list()
  list1$sAML1_0<- tt$sAML1$fistr_half$hgnc_symbol
  list1$AML4_0<- tt$AML4$fistr_half$hgnc_symbol
  list1$AML5_0<- tt$AML5$fistr_half$hgnc_symbol
com_gene_0<- Reduce(intersect,list1)%>%as.data.frame()
geneDescriptions_0 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_0, mart = ensembl)

####
list2=list()
  list2$sAML1_1<- tt$sAML1$seconed_half$hgnc_symbol
  list2$AML4_1<- tt$AML4$seconed_half$hgnc_symbol
  list2$AML5_1<- tt$AML5$seconed_half$hgnc_symbol
com_gene_1<- Reduce(intersect,list2)
geneDescriptions_1 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_1, mart = ensembl)

  ####

com_1_0<-intersect(com_gene_0,com_gene_1)

desc<- grch38 %>% filter(symbol %in% com_gene_1) %>% dplyr::select(ensgene, symbol, chr, start, end, description)

# edo <- DOSE::enrichDGN(G_list$entrezgene_id)
#   q<-dotplot(edo, showCategory=20) + labs(title = "Dotplot of the enriched pathways in NCG",subtitle =paste0(i,"_",p,"_NCG"))+
#   theme_bw() +theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 10),axis.text.y = element_text( size = 12))
# plot(q)
# ggsave(paste0(results_motif,i,"_",p,"_NCG",".jpg"),plot = q,width = 15,height = 15)
# edox <- setReadable(edo, 'org.Hs.eg.db', 'ENTREZID')
# p1 <- cnetplot(edox, foldChange=x)+  labs(subtitle = "Network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# p2 <- cnetplot(edox, foldChange=x, circular = TRUE, colorEdge = TRUE) +labs(subtitle = "Circular network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# q<-cowplot::plot_grid(p1, p2, ncol=2, labels=LETTERS[1:2],label_size = 1)
# ggsave(paste0(results_motif,i,"_",p,"_NCG_network",".jpg"),plot = q,width = 20,height = 10)
# 
# p1 <- heatplot(edox, showCategory=20)
# p2 <- heatplot(edox, foldChange=edo, showCategory=20)
# q<-cowplot::plot_grid(p1, p2, ncol=1, labels=LETTERS[1:2])
# ggsave(paste0(results_motif,i,"_",p,"_NCG_heatmap",".jpg"),plot = q,width = 20,height = 15)
# 



library(XML)

###########. RNA
results_motif<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/hBM1/results_RNA/"
dir.create(results_motif)
hsGO <- godata('org.Hs.eg.db', ont="MF")
ensembl=useMart("ensembl")
ensembl_d = useDataset("hsapiens_gene_ensembl",mart=ensembl)

tt<-lapply(samples,function(i){
  t<- lapply(c(0,1),function(p){
 xml_file_path <- paste0("/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/hBM1/Tomom_RNA_motif_enrichment_comparison/",i,"_",p,".xml")
#xml_file_path<-"/Users/ieo5268/Documents/Thesis_Iman/tomtom.xml"
# Parse the XML file
xml_data <- xmlParse(xml_file_path)

# Access elements and attributes in the XML file
#x <- getNodeSet(xml_data, "//motif")
motif_elements <- getNodeSet(xml_data, "//motif")
x <- sapply(motif_elements, function(motif) {
xmlGetAttr(motif, "alt")
})
  #x<-motif_data
  ##############
  gostres <- gost(query = x, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)
  
  p1<- gostplot(gostres, capped = T, interactive = F)
  p1<- p1+theme(text = element_text(size = 10))
  p1
  ggsave(paste0(results_motif,i,"_",p,"_gosplot_.","jpg"),plot = p1,width = 15,height = 7)
  publish_gosttable(gostres, highlight_terms = gostres$result[gostres$result$highlighted==T,]$term_id,
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size", "intersection_size"),
                        filename = paste0(results_motif,i,"_",p,"_gostable_.","pdf"))
  ############
  
  G_list <- getBM(attributes = c('entrezgene_id','hgnc_symbol',"ensembl_gene_id"), filters = 'hgnc_symbol', values =x, mart = ensembl_d)
  enr_bp<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "BP",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_cc<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "CC",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  enr_mf<-enrichGO(gene = G_list$entrezgene_id, OrgDb= org.Hs.eg.db, ont = "MF",pAdjustMethod = "BH",pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable= TRUE)
  a1<- barplot(enr_cc)+ labs(subtitle = paste0("Enriched_GO_","CC"))
  a2<- barplot(enr_bp)+ labs(subtitle = paste0("Enriched_GO_","BP"))
  a3<- barplot(enr_mf)+ labs(subtitle = paste0("Enriched_GO_","MF"))

  q<-cowplot::plot_grid( a1,a2,a3, ncol=3)
  ggsave(paste0(results_motif,i,"_",p,"_enriched_GO_123_",".jpg"),plot = q,width = 30,height = 7)
  
  kk <- enrichKEGG(gene = G_list$entrezgene_id,organism   = 'hsa', pvalueCutoff = 0.05)
  edox <- setReadable(kk, 'org.Hs.eg.db', 'ENTREZID')
  a1<- dotplot(edox)
  a2<- cnetplot(edox, categorySize="geneNum")
  q<-cowplot::plot_grid( a1,a2, ncol=2)
  ggsave(paste0(results_motif,i,"_",p,"_KEGG_",".jpg"),plot = q,width = 18,height = 7)
  
  pdf(paste0(results_motif,i,"_",p,"_enrichedGO_BP_",".pdf"),width = 12,height = 12)
  plotGOgraph(enr_bp)
  dev.off()
  
  
return(G_list)
  })
  names(t)<- c("fistr_half","seconed_half")
  return(t)
}
)
names(tt)<- samples
saveRDS(object = tt,file = paste0(results_motif,"tomtom_object_RNA.rds"))


# edo <- DOSE::enrichDGN(G_list$entrezgene_id)
#   q<-dotplot(edo, showCategory=20) + labs(title = "Dotplot of the enriched pathways in NCG",subtitle =paste0(i,"_",p,"_NCG"))+
#   theme_bw() +theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 10),axis.text.y = element_text( size = 12))
# plot(q)
# ggsave(paste0(results_motif,i,"_",p,"_NCG",".jpg"),plot = q,width = 15,height = 15)
# edox <- setReadable(edo, 'org.Hs.eg.db', 'ENTREZID')
# p1 <- cnetplot(edox, foldChange=x)+  labs(subtitle = "Network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# p2 <- cnetplot(edox, foldChange=x, circular = TRUE, colorEdge = TRUE) +labs(subtitle = "Circular network of the enriched pathways in NCG",caption =paste0(i,"_",p,"_NCG"))
# q<-cowplot::plot_grid(p1, p2, ncol=2, labels=LETTERS[1:2],label_size = 1)
# ggsave(paste0(results_motif,i,"_",p,"_NCG_network",".jpg"),plot = q,width = 20,height = 10)
# 
# p1 <- heatplot(edox, showCategory=20)
# p2 <- heatplot(edox, foldChange=edo, showCategory=20)
# q<-cowplot::plot_grid(p1, p2, ncol=1, labels=LETTERS[1:2])
# ggsave(paste0(results_motif,i,"_",p,"_NCG_heatmap",".jpg"),plot = q,width = 20,height = 15)
# 


```
## Intersection
```{r}
ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
DNA_mut_wt<-"/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/DNA_binding_sites/tomtom_object_DNA.rds"
DNA_mut_hbm<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/hBM1/results_DNA/tomtom_object_DNA.rds"
  
RNA_mut_wt<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/RNA_binding_sites/tomtom_object_RNA.rds"
RNA_mut_hbm<- "/Users/ieo5268/Documents/Thesis_Iman/cluster/motif_discovery/Motif_analysis/3rd_try/hBM1/results_RNA/tomtom_object_RNA.rds"

DNA_mut_wt_l<- readRDS(DNA_mut_wt)
DNA_mut_hbm_l<- readRDS(DNA_mut_hbm)
RNA_mut_wt_l<- readRDS(RNA_mut_wt)
RNA_mut_hbm_l<- readRDS(RNA_mut_hbm)


com_gene_0<- Reduce(intersect,DNA_mut_wt_l$sAML1)%>%as.data.frame()
geneDescriptions_0 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_0, mart = ensembl)

####
list2=list()
  list2$sAML1_1<- tt$sAML1$seconed_half$hgnc_symbol
  list2$AML4_1<- tt$AML4$seconed_half$hgnc_symbol
  list2$AML5_1<- tt$AML5$seconed_half$hgnc_symbol
com_gene_1<- Reduce(intersect,list2)
geneDescriptions_1 <- getBM(attributes = c("hgnc_symbol",'entrezgene_id', "description","wikigene_description"), filters = "hgnc_symbol", values = com_gene_1, mart = ensembl)

  ####

com_1_0<-intersect(geneDescriptions_1$hgnc_symbol,geneDescriptions_0$hgnc_symbol)

desc<- grch38 %>% filter(symbol %in% com_gene_1) %>% dplyr::select(ensgene, symbol, chr, start, end, description)
```








